shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/EpochBoundary.hs

```diff
@@ -219,46 +240,222 @@ instance
+emptySnapShots :: SnapShots era
+emptySnapShots = SnapShots (Completed emptySnapShot) emptySnapShot emptySnapShot (Coin 0)
+
+-- ===================================================================
+-- Pulsing Stake Distribution
+-- ===================================================================
+
+-- | A pulsing implementation of a SnapShot
+data PulsingStakeDistr era m
+  = StillPulsing !(StakeDistrPulser era m (SnapShot (Crypto era))) -- Pulsing work still to do
+  | Completed !(SnapShot (Crypto era)) -- Pulsing work completed, ultimate goal reached
+
+-- | Run a PulsingStakeDistr (a lazy SnapShot) until it is completed.
+runToCompletion ::
+  Era era =>
+  PulsingStakeDistr era Identity ->
+  SnapShot (Crypto era)
+runToCompletion (Completed ss) = ss
+runToCompletion (StillPulsing sdp) = runIdentity (completeM sdp)
+
+pulse ::
+  Era era =>
+  PulsingStakeDistr era Identity ->
+  PulsingStakeDistr era Identity
+pulse (Completed ss) = (Completed ss)
+pulse (StillPulsing sdp) =
+  if (done new)
+    then Completed (current new)
+    else StillPulsing new
+  where
+    new = (runIdentity (pulseM sdp))
+
+-- | We write our own Eq instance which completes one of the arguments when their constructors differ.
+instance
+  ( Era era,
+    Eq (UTxO era)
+  ) =>
+  Eq (PulsingStakeDistr era Identity)
+  where
+  (StillPulsing sdp1) == (StillPulsing sdp2) = sdp1 == sdp2
+  (Completed ss1) == (Completed ss2) = ss1 == ss2
+  (Completed ss1) == (StillPulsing sdp2) = ss1 == (runIdentity (completeM sdp2))
+  (StillPulsing sdp1) == (Completed ss2) = ss2 == (runIdentity (completeM sdp1))
+
+deriving instance (Show (UTxO era)) => Show (PulsingStakeDistr era m)
+
+deriving instance Generic (PulsingStakeDistr era m)
+
+instance (Era era, NoThunks (UTxO era)) => NoThunks (PulsingStakeDistr era m) where
+  showTypeOf _ = "PulsingStakeDistr"
+  wNoThunks ctxt (StillPulsing x) = allNoThunks [noThunks ctxt x]
+  wNoThunks ctxt (Completed x) = allNoThunks [noThunks ctxt x]
+
+instance (Era era, NFData (UTxO era)) => NFData (PulsingStakeDistr era m) where
+  rnf (StillPulsing x) = (rnf x)
+  rnf (Completed x) = (rnf x)
+
+instance (Typeable m, Era era, ToCBOR (UTxO era)) => ToCBOR (PulsingStakeDistr era m) where
+  toCBOR (StillPulsing p) = encode (Sum (StillPulsing @era) 0 !> To p)
+  toCBOR (Completed r) = encode (Sum (Completed @era) 1 !> To r)
+
+instance (Typeable m, Era era, FromCBOR (UTxO era), FromCBOR (Core.TxOut era)) => FromCBOR (PulsingStakeDistr era m) where
+  fromCBOR = decode (Summands "PulsingStakeDistr" decPSD)
+    where
+      decPSD 0 = (SumD StillPulsing <! From)
+      decPSD 1 = (SumD Completed <! From)
+      decPSD n = Invalid n
+
+-- ===============================================================================
+
+-- | StakeDistrPulser acts like a serializable closure for the lazy computation of a SnapShot.
+data StakeDistrPulser era (m :: Type -> Type) ans where
```

Comment:
Can we document the parameters here? What is `m`?

Reply:
What's the reason for abstracting over `ans` when it's then fixed by the only constructor?

Reply:
So StakeDistrPulser (like all puslers) is a monad transformer, that runs with the underlying monad 'm', The methods of the Pulsable class abstract over the "ans" type (i.e what you get when pulsing is complete. So the type that is a member of the Pulsable class must be a Ternary type data structure.
   class Pulsable (pulse :: (Type -> Type) -> Type -> Type) where ...
In the current instance the "ans" type is completely concrete. But  StakeDistrPulser  still must be ternary. This way we get to use all the Pulsable structure and methods and morphisms.

Reply:
I can certainly document this better.

Reply:
StakeDistrPulser acts like a serializable closure for the lazy computation of a SnapShot.
It will be made an instance of Puslsable, so it need to be a Ternary type constuctor.
'm' is the monad that the pulsing action is executed in, and 'ans' is the type of the object
computed by the closure. (it will be SnapShot). We could inline 'ans' to SnapShot, but then we could not make a Pulsable instance for it.

I added comment to this effect in the code.